src/core/algorithms/gfd/gfd_miner.cpp

```diff
@@ -0,0 +1,641 @@
+    }
+}
+
+bool GfdMiner::Satisfied(std::vector<Literal> const& literals, graph_t const& pattern,
+                         Embedding const& embedding) {
+    auto check_literal = [&](auto const& l) {
+        auto const& [fst_token, snd_token] = l;
+        std::string fst, snd;
+        bool first_res = CheckToken(fst_token, fst, pattern, embedding);
+        bool second_res = CheckToken(snd_token, snd, pattern, embedding);
+        return first_res && second_res && (fst == snd);
+    };
+    return std::ranges::all_of(literals, check_literal);
+}
+
+void GfdMiner::AddCompacted(std::vector<SingleRule>& rules, graph_t& pattern,
+                            Rules& forbidden_rules) {
+    if (rules.empty()) return;
+    std::ranges::sort(rules);
+    std::vector<Literal> prev = rules.front().first;
+    std::vector<Literal> conclusion = {};
+    for (auto const& [premises, l] : rules) {
+        if (prev != premises) {
+            gfds_.emplace_back(pattern, prev, conclusion);
+            forbidden_rules.emplace_back(std::move(prev), std::move(conclusion));
+            prev = premises;
+            conclusion = {l};
+        } else {
+            conclusion.push_back(l);
+        }
+    }
+    gfds_.emplace_back(pattern, prev, conclusion);
+    forbidden_rules.emplace_back(std::move(prev), std::move(conclusion));
+}
+
+bool GfdMiner::Validate(Gfd const& gfd, Embeddings const& embeddings) {
+    bool result = false;
+    VertexImage vertex_map;
+
+    for (Embedding const& embedding : embeddings) {
+        if (!Satisfied(gfd.GetPremises(), gfd.GetPattern(), embedding)) {
+            continue;
+        }
+        result = true;
+
+        if (!Satisfied(gfd.GetConclusion(), gfd.GetPattern(), embedding)) {
+            return false;
+        }
+
+        PopulateVertexMap(vertex_map, gfd.GetPattern(), embedding);
+    }
+
+    std::size_t min_size = GetMinSize(vertex_map, boost::num_vertices(graph_));
+    return result && (min_size >= sigma_);
+}
+
+void GfdMiner::HGenerate(std::vector<Literal>& notsatisfied, Literal const& l,
+                         std::vector<SingleRule>& rules, graph_t& pattern,
+                         Embeddings const& embeddings, Rules const& forbidden_rules) {
+    std::vector<Literal> conclusion{l};
+    for (std::size_t k = 1; k <= 3; ++k) {
+        std::vector<std::vector<Literal>> subsets = GetSubsets<Literal>(notsatisfied, k);
+        for (auto& subset : subsets) {
+            if (IsForbidden(forbidden_rules, subset, l)) {
+                continue;
+            }
+
+            Gfd gfd = {pattern, subset, conclusion};
+            if (Validate(gfd, embeddings)) {
+                rules.emplace_back(subset, l);
+                auto cond = [&subset](Literal const& lit) {
+                    return std::ranges::find(subset, lit) != subset.end();
+                };
+                auto it = std::ranges::remove_if(notsatisfied, cond).begin();
+                notsatisfied.erase(it, notsatisfied.end());
```

Comment:
use `std::erase_if`:
```suggestion
                std::erase_if(notsatisfied, cond);
```